app/controllers/wishlist_controller.rb

```diff
@@ -1,6 +1,57 @@
 class WishlistController < ApplicationController
+  before_action :authenticate_user!
+  before_action :find_wishlist_item
 
   # GET /wishlist
   def index
+    @wishlist = current_user.wishlist
+  end
+
+  # POST /wishlist
+  def add_course
+    respond_to do |format|
+      if @wishlist_item.save
+        flash[:notice] = 'Course successfully saved to your wishlist'
+        format.html { redirect_to params[:url] || wishlist_path }
+      else
+        flash[:notice] = 'An error occurred adding the course to your wishlist'
+        format.html { redirect_to params[:url] || wishlist_path }
+      end
+    end
+  end
+
+  # PUT /wishlist
+  def update
+    respond_to do |format|
+      if @wishlist_item.update(wishlist_item_params)
+        format.html { redirect_to wishlist_path }
+      else
+        format.html { redirect_to wishlist_path }
+      end
+    end
+  end
+
+  # DELETE /wishlist
+  def remove_course
+    respond_to do |format|
+      if @wishlist_item.destroy
+        flash[:notice] = 'Course successfully removed from your wishlist'
+        format.html { redirect_to params[:url] || wishlist_path }
+      else
+        flash[:notice] = 'An error occurred while attempting to remove the course from your wishlist'
+        format.html { redirect_to params[:url] || wishlist_path }
+      end
+    end
+  end
+
+  private
+
+  def find_wishlist_item
+    @wishlist_item = WishlistItem.where(user: current_user, course_id: params[:course_id])
+                                 .first_or_initialize
```

Comment:
We are calling this method on every page in the controller, is this correct? Seems like we shouldn't have an `@wishlsit_item` when we're viewing the index page.